### ISSUE TITLE
Type narrowing doesn't work correctly if in a loop with lambda

### ISSUE THREAD

**Issue by PRESIDENT810** [microsoft/pyright#9838]
## Environment data

-   Pylance version: 2025.2.1
-   OS and version: macOS 14.7.2
-   Python version (& distribution if applicable, e.g. Anaconda): miniconda, Python 3.12.4

## Code Snippet

```python
from typing import Union, Any

class Unset:
    pass

def map_to_dict(optional: str) -> dict[str, Any]:
        return {}

def generate_optional(i: int) -> Union[Unset, str]:
    return "" if i == 0 else Unset()

def foo1(i: int):
    optional = generate_optional(i)
    optional = "" if isinstance(optional, Unset) else optional
    """
    Pylance is OK with this
    """
    something = map(lambda x: map_to_dict(optional), [])    

def foo2():
    for i in range(3):
        optional = generate_optional(i)
        optional = "" if isinstance(optional, Unset) else optional
        """
        Pylance complains:
        Argument of type "Unset | str" cannot be assigned to parameter "optional" of type "str" in function "map_to_dict"
          Type "Unset | str" is not assignable to type "str"
        "Unset" is not assignable to "str"PylancereportArgumentType
        """
        something = map(lambda x: map_to_dict(optional), [])    
```

## Repro Steps

Just try the code snippet above and you will have

<img width="1231" alt="Image" src="https://github.com/user-attachments/assets/6b43129b-a238-4ff7-a84b-bc7c2f9491bd" />

I think it's clear that in `foo2`, the scope of `optional` should be within the for loop, and you can see there's no subsequent assignment that might change the type of `optional`, so pylance should be able to narrow its type to `str` (actually if I hover my cursor to `optional` in line 23 it says the type is `str`)

<img width="552" alt="Image" src="https://github.com/user-attachments/assets/b6b8160d-6459-4202-b5b6-ce43c729502c" />

However in line 30 pylance still complains the type might be Unset.

`foo2` is just `foo1` within a loop, so I think if `foo1` is able to narrow the type to `str` correctly, `foo2` should be able to do the same?

```
XXX
```

**Comment by erictraut** [comment]
This is the intended behavior. In general, it's unsafe for a type checker to retain a narrowed type for an inner scope like a lambda because the value of captured values can be modified in other scopes. Pyright (the type checker that underlies pylance) is able to retain narrowed types under certain cases where it can prove that it's safe to do so. The presence of a loop defeats this check. For more details, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=narrowing-for-captured-variables).

You can work around this limitation within your code by assigning the narrowed value to a temporary variable that is then captured.

```python
def foo2():
    for i in range(3):
        optional = generate_optional(i)
        optional = "" if isinstance(optional, Unset) else optional
        temp = optional
        something = map(lambda x: map_to_dict(temp), [])
```
